Add bf16 gemm pingpong with num_stages=3

jungpark-mlir opened the pull request:
Transforms 'BLOCK_SIZE_M': 256, 'BLOCK_SIZE_N': 256, 'BLOCK_SIZE_K': 32 with num_stages=3.
With tripke buffering, it can deal hazard between waves more easily. e.g., wait/copy/dot can be freely ordered.
This requires carrying wait token over the loop (TRITON_HIP_ENABLE_F16_ASYNC_PINGPONG=1).
Waits are merged in the pingpong pass for now, since pipeliner is not able to merge it correctly.

jungpark-mlir committed:
Add bf16 gemm pingpong with num_stages=3

Transforms 'BLOCK_SIZE_M': 256, 'BLOCK_SIZE_N': 256, 'BLOCK_SIZE_K': 32
with num_stages=3, it can deal hazard between waves more easily.
Requires carrying wait token over the loop, waits are merged in the
pingpong pass.

guacamoleo commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -1213,6 +1253,7 @@ void Pingponger::getDotPingponged() {
     return;
   }
 
+
```
nit: extra newline

guacamoleo commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -1126,6 +1125,49 @@ LogicalResult Pingponger::transformFP4(OpBuilder &builder, Location loc) {
   return success();
 }
 
+LogicalResult Pingponger::transformNS3(OpBuilder &builder, Location loc) {
```
I see there's already a comment for documentation; but it would be great to capture what these transform functions do; e.g. what is the assumed state of the IR before tranformation, how/when is conditional barrier inserted, what the execution mode looks like before/after the transformation.

guacamoleo commented on third_party/amd/lib/TritonAMDGPUTransforms/StreamPipeline.cpp:
```diff
@@ -1128,7 +1128,7 @@ struct PipelinePass : public TritonAMDGPUStreamPipelineBase<PipelinePass> {
     // So we should just not do it to get a better vmcnt in front of each
     // AsyncCopy.
-    if (useAsyncCopy) {
+    if (useAsyncCopy && !useF16BlockPingpong) {
```
Is this PR adding support for other precisions  to use numStages=3, or only 16bit is allowed to use num_stages=3?

jungpark-mlir replied:
This part is actually adding restriction to use `tt:combineRedundantWaitOps` which is problematic with changed ordering with f16 pingpong in the pipeliner, which carry over the tokens from async_wait, instead of carrying over tokens from async_commit.

guacamoleo commented on third_party/amd/backend/compiler.py:
```diff
@@ -266,7 +266,7 @@ def make_ttgir(mod, metadata, options):
         amd.passes.ttgpuir.add_reorder_instructions(pm)
         use_block_pingpong = is_pingpong_schedule_enabled(options.arch)
-        if use_block_pingpong and options.num_stages in [2, 4]:
+        if use_block_pingpong and options.num_stages in [2, 3, 4]:
```
Do we only allow pingpong to be enabled for pre-vetted scenarios where we know that pingpong is faster than non-pingpong? Or do we allow the user to enable pingpong even when it's slower (e.g. for testing or analysis...)?
I ask because I'm wondering if we distinguish between "supported" cases being numerically correct or performant also.

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -1302,11 +1343,19 @@ void Pingponger::getDotPingponged() {
       LDBG("Currently only support num_warp=8 for async PP");
       return;
     }
+    if (numStages == 3 && dotOps.size() == 1 && tileSize == mediumTile && aShape[1] == 32 && elemWidth == 16) {
```
@guacamoleo it's guarded here, only enabled with supported tile size (and expected to be performant) and data type. Checking num_stages=3 is just first step to check it.

jungpark-mlir committed:
Fine tuning NS=3 pingpong.

Try to reduce salu/valu stalls by
using sched.group.barrier to interleave MFMA and salu/buffer_load
using sched.barrier to split ds_read and interleave address calcs.
